### ISSUE TITLE
[HOLD for payment 2023-11-22] [$500]  Group chat - Room name in settings is different than the name in chat report and LHN title

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#30257]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.90-1
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

**Issue found when executing PR** https://github.com/Expensify/App/pull/28991
## Action Performed:
1. Go to staging.new.expensify.com
2. Create a group chat with at least four members
3. Click on the chat header
4. Go to Settings
Note that the room name is different from the name in LHN and chat header

## Expected Result:
The room name is consistent with the name in the group chat header and report title in LHN

## Actual Result:
The room name is different from the name in the group chat header and report title in LHN

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

 

https://github.com/Expensify/App/assets/78819774/d0b0eae5-5494-4263-b5f4-0b1af1058e44

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010327954e256575ce</li>
        <li>Upwork Job ID: 1716822959251263488</li>
        <li>Last Price Increase: 2023-10-24</li>
<li>Automatic offers: </li>
<ul>
<li>DylanDylann | Contributor | 27544084</li>
</ul></li>
    </ul>
</details>

**Comment by esh-g** [comment]
I think the room name is set the backend and the sorting should take place there like it does in `ReportUtils.getDisplayNamesWithTooltips`

**Comment by akamefi202** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Room name in settings is different than the name in chat report and LHN title.

### What is the root cause of that problem?
https://github.com/Expensify/App/blob/485a851cf9068e0cc25df690b1652e078455372a/src/libs/ReportUtils.js#L2044

Room name in settings gets the value from `getReportName` function without sorting.

https://github.com/Expensify/App/blob/485a851cf9068e0cc25df690b1652e078455372a/src/libs/ReportUtils.js#L1231-L1236

But other room names get the value from `getDisplayNamesWithTooltips` function with sorting.
Sorting makes difference.

### What changes do you think we should make in order to solve the problem?
We can add sorting before returning value in `getReportName` function too.

### What alternative solutions did you explore? (Optional)
N/A

**Comment by syedsaroshfarrukhdot** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Room name in settings is different than the name in chat report and LHN title.

### What is the root cause of that problem?

Room Name in getReportName Function is not returning sorted list.

https://github.com/Expensify/App/blob/485a851cf9068e0cc25df690b1652e078455372a/src/libs/ReportUtils.js#L2044

### What changes do you think we should make in order to solve the problem?


We need to return sortedParticipants by sorting Participants in same way they are being sorted  in getDisplayNamesWithTooltips

           const sortedParticipants = _.chain(participantsWithoutCurrentUser)
                 .map((accountID) => getDisplayNameForParticipant(accountID, isMultipleParticipantReport))
                 .sort((first, second) => first.localeCompare(second))
                 .value();

           return sortedParticipants.join(', ');

### What alternative solutions did you explore? (Optional)

N/A

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- Group chat - Room name in settings is different than the name in chat report and LHN title

### What is the root cause of that problem?
- The ways we get the room nane in settings is different than the way we did in LHNRow and Report `s header.
- Here is how we get the room name in settings:
https://github.com/Expensify/App/blob/414d5913999f69d83ab08b6ae6e03325c9ede549/src/pages/settings/Report/ReportSettingsPage.js#L81
- Here is how we did with Report `s header and same with LHNRow:
https://github.com/Expensify/App/blob/414d5913999f69d83ab08b6ae6e03325c9ede549/src/pages/home/HeaderView.js#L85
### What changes do you think we should make in order to solve the problem?
- We should create the util function like ````getGroupChatName``` that based on what we did with LHNRow and Report `s header and use in these above cases like:
```
const getGroupChatName = (report) => {
    const participants = lodashGet(report, 'participantAccountIDs', []);
    const isMultipleParticipantReport = participants.length > 1;
    const participantPersonalDetails = OptionsListUtils.getPersonalDetailsForAccountIDs(participants, allPersonalDetails);
    const displayNamesWithTooltips = getDisplayNamesWithTooltips(participantPersonalDetails, isMultipleParticipantReport);
    return getDisplayNamesStringFromTooltips(displayNamesWithTooltips);
};
```

### What alternative solutions did you explore? (Optional)
- NA

**Comment by burczu** [comment]
not overdue - I'll review proposals soon

**Comment by burczu** [comment]
I've reviewed all the proposals and I think all contributors identified the root cause correctly.

I like the most [the proposal](https://github.com/Expensify/App/issues/30257#issuecomment-1780885889) from @DylanDylann because I believe we should not change the implementation of the `getReportName` function itself as it's used in many places in the app, and sorting may not be expected in other places.

🎀 👀 🎀 C+ reviewed

**Comment by akamefi202** [comment]
@burczu Thank you for your review. But I think that we don't need to add a new function for this.

https://github.com/Expensify/App/blob/f8e77debf5a9351c2b377e72e850da7837961318/src/libs/ReportUtils.js#L4099-L4109

This is how `isGroupChat` function works and it is calling other functions such as `isChatThread`, `isTaskReport`, etc.
And similar logic already exists inside `getReportName` function.

https://github.com/Expensify/App/blob/f8e77debf5a9351c2b377e72e850da7837961318/src/libs/ReportUtils.js#L2045-L2054

`getReportName` function is also calling functions such as `isChatThread`, `isTaskReport`, etc for categorization.
And it will return `formattedName` if the report is not group chat.

So the code part that I suggested to add sorting will not be called unless the report is group chat.
That's why I think that we don't need to add a separate function for group chat.
Thank you for consideration. And please share you thoughts.

**Comment by burczu** [comment]
@anmurali Just to inform, I'll be ooo for the next few days and we'll be back on Monday, November 6th.

**Comment by DylanDylann** [comment]
@Li357 pleae help review the proposals as C+ `ve reviewed

**Comment by Li357** [comment]
Agree with deduplicating logic here. @DylanDylann all yours

**Comment by DylanDylann** [comment]
@burczu PR https://github.com/Expensify/App/pull/31016 is ready for review

**Comment by burczu** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

* [x]  [@burczu] The PR that introduced the bug has been identified. Link to the PR:

**I don't think this issue is an regression caused by another PR - room name settings, chat report name and LHN title was developed separately and this discrepancy was just missed.**

* [x]  [@burczu] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

**n/a**

* [x]  [@burczu] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

**n/a**

* [x]  [@burczu] Determine if we should create a regression test for this bug.

**I think this is not very impactul issue, and isn't related to any super important flow, so I tend to believe that we don't need to add regression tests here. Also, in the PR fixing this issue, we extracted providing the group chat name to the separate method and now using it in all the affected places, so I think this not very likely this issue will occur again.**

* [x]  [@burczu] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
* [ ]  [@anmurali] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by anmurali** [comment]
Paid!
